xla/service/gpu/cudnn_fusion_compiler.cc

```diff
@@ -0,0 +1,430 @@
+    if (absl::c_any_of(dot->precision_config().operand_precision(),
+                       [](int x) { return x != PrecisionConfig::DEFAULT; })) {
+      VLOG(3) << "Non-default precision is not supported.";
+      return std::nullopt;
+    }
+    TF_ASSIGN_OR_RETURN(auto analysis,
+                        TritonFusionAnalysis::Execute(computation));
+    return GemmDimensionAdapter{*dot, std::move(analysis)};
+  }
+
+  bool DimensionsAndStrides(const HloInstruction& hlo,
+                            const TritonFusionAnalysis::Scope scope,
+                            std::vector<int64_t>& dimensions,
+                            std::vector<int64_t>& strides) {
+    const DotDimensionNumbers& dims = dot_.dot_dimension_numbers();
+    // GEMM fusions require a specific canonical order of dimensions.
+    std::vector<int64_t> dim_indices;
+    switch (scope) {
+      case TritonFusionAnalysis::Scope::LHS:
+        dim_indices = {dims.lhs_batch_dimensions().empty()
+                           ? -1
+                           : dims.lhs_batch_dimensions(0),
+                       GetNonContractingDims(dot_.operand(0)->shape(),
+                                             dims.lhs_batch_dimensions(),
+                                             dims.lhs_contracting_dimensions())
+                           .value()[0],
+                       dims.lhs_contracting_dimensions(0)};
+        break;
+      case TritonFusionAnalysis::Scope::RHS:
+        dim_indices = {dims.rhs_batch_dimensions().empty()
+                           ? -1
+                           : dims.rhs_batch_dimensions(0),
+                       dims.rhs_contracting_dimensions(0),
+                       GetNonContractingDims(dot_.operand(1)->shape(),
+                                             dims.rhs_batch_dimensions(),
+                                             dims.rhs_contracting_dimensions())
+                           .value()[0]};
+        break;
+      case TritonFusionAnalysis::Scope::OUTPUT:
+        dim_indices = {dims.lhs_batch_dimensions().empty() ? -1 : 0,
+                       dot_.shape().rank() - 2, dot_.shape().rank() - 1};
+        break;
+    }
+    dimensions.reserve(dim_indices.size());
+    strides.reserve(dim_indices.size());
+    for (const int index : dim_indices) {
+      const auto* spec = analysis_.IterSpec(scope, &hlo, index);
+      if (spec == nullptr) {
+        dimensions.push_back(1);
+        strides.push_back(strides.empty() ? 1 : strides.back());
+        continue;
+      } else {
+        if (spec->size() != 1) {
+          return false;
+        }
+        dimensions.push_back(spec->front().count);
+        strides.push_back(spec->front().stride);
+      }
+    }
+    return true;
+  }
+
+ private:
+  const HloDotInstruction& dot_;
+};
+
+// Traverses fusion computations and creates cuDNN graphs out of them.
+StatusOr<std::optional<graph::Graph>> HloFusionToCuDnnGraph(
```

Comment:
Can you please add some tests for the translation logic HLO->cuDNN graphs? Looks like this is only indirectly tested via cudnn_test.cc so far. Seeing HLO next to the expected cuDNN graphs in tests would be really helpful to those changing this file or just looking to understand the logic better.

Reply:
cuDNN graphs do not have text-only serialization right now, the one that is currently in use includes compiled binaries. I forwarded the request, it should become available very soon.